Kubernetes Tutorial 1 Intro (#187)
ippontech/blog-usa: posts/kubernetes-tutorial-1-intro.md

---
authors:
- David Elizondo
tags:
- Kubernetes
date: 2021-09-07T12:21:50.000Z

> title: "Kubernetes Tutorial 1 Intro"

Reviewer: ```suggestion
title: "Kubernetes Tutorial #1: Intro"
```

image: 
---

## Intro 

> Devops Engineers are now dealing with large application systems in the cloud. One of the most popular orchestration systems currently being used in production across the board is Kubernetes (K8s) under the covers. I want to share some knowledge I have gathered over the years and break down what k8s does at it's core. Understanding of the K8s core construction will allow your to understand how new features are constructed for k8s and how these basic functions ensure a more stable operating environment. 

Reviewer: ```suggestion
DevOps Engineers are now dealing with large application systems in the cloud. One of the most popular orchestration systems currently being used in production across the board is Kubernetes (K8s, for short) under the covers. I want to share some knowledge I have gathered over the years and break down what K8s does at its core. An understanding of the K8s core construction will allow you to understand how new features are constructed for K8s and how these basic functions ensure a more stable operating environment. 
```

> Devops Engineers are now dealing with large application systems in the cloud. One of the most popular orchestration systems currently being used in production across the board is Kubernetes (K8s) under the covers. I want to share some knowledge I have gathered over the years and break down what k8s does at it's core. Understanding of the K8s core construction will allow your to understand how new features are constructed for k8s and how these basic functions ensure a more stable operating environment. 

Reviewer 2: ```suggestion
DevOps engineers have moved from on-premise, monolithic applications to large-scale application systems in the cloud. One of the most popular orchestration systems currently being used in production across the board is Kubernetes (K8s). I want to share some knowledge I have gathered over the years and break down what K8s does at its core. Understanding of the K8s core construction will allow you to understand how new features are constructed for K8s and how these basic functions ensure a more stable operating environment. 
```

> Kubernetes is an open source project on [github](https://github.com/kubernetes/kubernetes). They define their project as: 

Reviewer 2: ```suggestion
[Kubernetes](https://kubernetes.io/) is an open-source project on [GitHub](https://github.com/kubernetes/kubernetes). They define their project as: 
```

> Kubernetes, also known as K8s, is an open source system for managing containerized applications across multiple hosts. It provides basic mechanisms for deployment, maintenance, and scaling of applications.

## Application Deployment

> To understand K8s, I believe it's helpful to look at and compare how we deployed production applications in the past and what k8s brings to the table. First lets consider the our first old school setup. At company ABC there is a couple of servers in the basement for hosting our applications. Lets say we size our production servers to match our expected user base and concurrent usage. In our example we have four business applications (Green,Orange,Teal,Blue). Lets say all these applications share a network, database and load balancer with each other. We also have a couple of spare servers allocated in the event of a disaster recovery.

Reviewer: ```suggestion
To understand Kubernetes, I believe it is helpful to look at and compare how production applications have been deployed in the past and compare that what K8s brings to the table in the present. First, let's consider our first "old school" setup. At company "ABC," there are a couple of servers in the basement to host our applications. Let's say we size our production servers to match our expected user base and concurrent usage. In our example, we have four business applications (Green, Orange, Teal, and Blue). Now let's say all of these applications share a network, database, and load balancer with each other. We also have a couple of spare servers allocated in the event of a disaster recovery.
```

> To understand K8s, I believe it's helpful to look at and compare how we deployed production applications in the past and what k8s brings to the table. First lets consider the our first old school setup. At company ABC there is a couple of servers in the basement for hosting our applications. Lets say we size our production servers to match our expected user base and concurrent usage. In our example we have four business applications (Green,Orange,Teal,Blue). Lets say all these applications share a network, database and load balancer with each other. We also have a couple of spare servers allocated in the event of a disaster recovery.

Reviewer 2: ```suggestion
To understand K8s, I believe it's helpful to look at and compare how we deployed production applications in the past and what K8s brings to the table. First let's consider our first legacy setup. At company ABC there are a couple of servers in a physical data center on the company's real estate somewhere for hosting our applications. Let's say we size our production servers to match our expected user base and concurrent usage. In our example we have four discrete business applications: Green, Orange, Teal, and Blue). Let's say all these applications share a network, database, and load balancer with each other. We also have a couple of spare servers allocated in the event of a disaster recovery.
```

![onprem-deployment](https://github.com/David-Elizondo/blog-usa/blob/master/images/2021/09/onprem.png)

## Simulating a Disaster

> Lets say the IT get alerted about an outage of application Blue. An electrical storm has come by that night and shorted the hardware on the server. During the outage IT determines that they will need to move the application to one of the spare servers. 

Reviewer: ```suggestion
We have a scenario where IT gets alerted about an outage of application "Blue." An electrical storm came through the night before and shorted the hardware on the server. Because of the outage, IT determines that they will need to move the application to one of the spare servers.
```

> Lets say the IT get alerted about an outage of application Blue. An electrical storm has come by that night and shorted the hardware on the server. During the outage IT determines that they will need to move the application to one of the spare servers. 

Reviewer 2: ```suggestion
Let's say the IT network operations center gets alerted about an outage in application Blue. An electrical storm has come by that night and shorted the hardware on the server. During the outage, IT determines that they will need to move the application to one of the spare servers. 
```

![onprem-deployment](https://github.com/David-Elizondo/blog-usa/blob/master/images/2021/09/onprem2.png)

> During the outage IT will be scrambling to bring one of the spare servers online. Configure access the shared database, network load balancer and allowing firewall rules for the application to run on the new server. Each of these task can be time consuming if not planned for. 

Reviewer: ```suggestion
During this outage, IT will be scrambling to bring one of the spare servers online. They must configure access to the shared database and network load balancer, and allow firewall rules for the application to run on the new server. Each of these tasks can be time-consuming if not planned for ahead of time. 
```

> During the outage IT will be scrambling to bring one of the spare servers online. Configure access the shared database, network load balancer and allowing firewall rules for the application to run on the new server. Each of these task can be time consuming if not planned for. 

Reviewer 2: ```suggestion
During the outage, IT will be scrambling to bring one of the spare servers online. In addition, they will have to configure access to the shared database, network load balancer, and allowing security system rules for the application to run on the new server. Each of these tasks can be time consuming if not planned for or have had repeated dry runs to ensure that they can be done issue-free.
```

> In our example each application can not recover from a single machine failure. Recovery from said disaster will bring outage times and IT time from investigation to recovery. Naturally we'd ask how can be setup our environment such that our application can recover from this scenario?

Reviewer: ```suggestion
In our example, each application can **not** recover from a single machine failure. Recovery from said disaster will bring outage times and IT time from investigation to recovery. Naturally, we would ask how can our environment be setup such that our application can recover from this scenario?
```

> In our example each application can not recover from a single machine failure. Recovery from said disaster will bring outage times and IT time from investigation to recovery. Naturally we'd ask how can be setup our environment such that our application can recover from this scenario?

Reviewer 2: ```suggestion
In our example each application cannot recover from a single machine failure. Recovery from said disaster will bring outage times and take IT time to perform investigation and recovery. Naturally, we'd ask how can be setup our environment such that our application can recover from this scenario?
```

> ## Break Down 

Reviewer: ```suggestion
## Breakdown 
```

> In our example losing one machine means having to manually recover. There are many ways to make your recovery time better in this scenario. For instance more equipment to protect against electrical storm or establish a process to build and test a DR plan in case of such disaster. These type of protections don't change the underlying issue which is single machine recovery. Another method used on-prem is to have one hot machine and a cold machine. The hot machine is used until a issue arises where the cold machine to switched to in case of backup. In this way now you can recover from a single failure but you've only moved the bar one tick and still would not be able to recover from a two machine failure. Ideally we wan't to be able to recovery from N machine failures. 

Reviewer: ```suggestion
Losing one machine means having to manually recover. There are many ways to make your recovery time more efficient in this scenario. For instance, install equipment to protect against electrical storm damage or establishing a process to build and test a DR plan in case of such disaster. These type of protections will not necessarily change the underlying issue, which is single machine recovery. Another method used on-prem is to have one hot machine and one cold machine. The hot machine is used until an issue arises, whereas the cold machine is switched on in case backup is necessary. You can now recover from a single failure, but you have only moved the bar one tick and still would not be able to recover from a two machine failure. Ideally, we want to be able to recover from any and all machine failures. 
```

> In our example losing one machine means having to manually recover. There are many ways to make your recovery time better in this scenario. For instance more equipment to protect against electrical storm or establish a process to build and test a DR plan in case of such disaster. These type of protections don't change the underlying issue which is single machine recovery. Another method used on-prem is to have one hot machine and a cold machine. The hot machine is used until a issue arises where the cold machine to switched to in case of backup. In this way now you can recover from a single failure but you've only moved the bar one tick and still would not be able to recover from a two machine failure. Ideally we wan't to be able to recovery from N machine failures. 

Reviewer 2: ```suggestion
In our example, losing one machine means having to manually recover. There are many ways to make your recovery time faster in this scenario. For instance, more equipment to protect against electrical storm or establish a process to build and test a DR plan in case of such disaster. These types of protections don't change the underlying issue which is single machine recovery. Another method used on-premises is to have one hot machine and a cold machine. The hot machine is used until an issue arises where the cold machine to switched to in case of backup. In this way now you can recover from a single failure, but you've only moved the bar one tick and still would not be able to recover from a two-machine failure. Ideally, we won't to be able to recover from an arbitrarily large amount of machine failures. 
```

> To achieve this K8s changes the way we deploy applications. Instead of having dedicated hardware and manual deployment processes. Our applications have automated deployments and containerized applications. 

Reviewer: ```suggestion
In order to achieve this, Kubernetes changes the way we deploy applications. Instead of having dedicated hardware and manual deployment processes, our applications now have automated deployments and containerized applications. 
```

> To achieve this K8s changes the way we deploy applications. Instead of having dedicated hardware and manual deployment processes. Our applications have automated deployments and containerized applications. 

Reviewer 2: This would be a good place to discuss the benefits of ephemeral infrastructure.

Author: What kind of benefits were you thinking of?

Reviewer 2: My thought was that if you are used to deploying applications where you are wiping the infrastructure and redeploying them from IaC, you become more resilient than deploying to a physical box in a datacenter or a static VM. Maybe this is out-of-scope for this. The DR talk made me think of this.

Author: K8s can be deployed both on-prem and in the cloud. So you can gain operational reliability by using orchestration from both perspectives. I think the benefits of IaC are mostly realized in the cloud,, So this topic would be better fit in my next blog where we take a dive into Distributions and Hosting on the cloud.

> [Kubernetes](https://kubernetes.io/) is one of many systems that enable this functionality by creating a distributed cluster of machines that all work together to host applications deployed. 

Reviewer: ```suggestion
[Kubernetes](https://kubernetes.io/) is one of many options that enable this type of functionality by creating a distributed cluster of machines that all work together to host deployed applications. 
```

> ## Cluster communication

Reviewer: ```suggestion
## Cluster Communication
```

> The way K8s operates is to have a pool of machines ready for any application to be deployed onto them. The machines communicate by having the same Kubelet application installed on all the machines. The Kubelet can manage the node it is installed on by running application containers and also has a communication endpoint for controllers nodes to send operations. 

Reviewer: ```suggestion
The way Kubernetes operates is to have a pool of machines ready for any application to be deployed onto them. The machines communicate by having the same Kubelet application installed on all said machines. The Kubelet can manage the node it is installed on by running application containers, while also having a communication endpoint for controllers nodes to send operations. 
```

> The way K8s operates is to have a pool of machines ready for any application to be deployed onto them. The machines communicate by having the same Kubelet application installed on all the machines. The Kubelet can manage the node it is installed on by running application containers and also has a communication endpoint for controllers nodes to send operations. 

Reviewer 2: Define Kubelet for the layman?

Author: Updated paragraph to include some explanation: 

>  The Kubelet has the responsibility of managing the node its installed on and communicating to the controller nodes. It has the system permissions to start and stop containers, mount and unmount volumes, and open and close ports etc. Essentially it is in charge of listening for commands sent from the controller nodes and executing the commands sent. This enables each node in the cluster to be orchestrated as a group of machines, otherwise known as a cluster.

> ## Visualize K8s 

Reviewer: ```suggestion
## Visualizing Kubernetes
```

> Back to Company ABC with four application deployments. We convert our production deployment operations to k8s. First by installing a Kubelet onto all of our agent nodes. 

Reviewer: ```suggestion
Let's go back to company "ABC" with their four application deployments. We convert our production deployment operations to Kubernetes -- first by installing a Kubelet onto all of our agent nodes.
```

> Back to Company ABC with four application deployments. We convert our production deployment operations to k8s. First by installing a Kubelet onto all of our agent nodes. 

Reviewer 2: ```suggestion
Back to Company ABC with four application deployments. We convert our production deployment operations to K8s, first by installing a Kubelet onto all our agent nodes. 
```

![k8s-deployment](https://github.com/David-Elizondo/blog-usa/blob/master/images/2021/09/k8s.png)

> Next we segment three machines to be controllers that will be in-charge of maintaining communications with all the agent nodes and checking health of the cluster configuration. 

Reviewer: ```suggestion
Next, we segment three machines to be controllers that will be in charge of maintaining communications with all the agent nodes and checking the health of the cluster configuration. 
```

![k8s-deployment](https://github.com/David-Elizondo/blog-usa/blob/master/images/2021/09/k8s2.png)

> We then define our application deployments as "Plans". The plans are defined by k8s configuration yaml files. These config files can configure many different aspects like resource sizes, network settings and health checks. These plans get communicated to our agent nodes for deployment on all active machines. 

Reviewer: ```suggestion
We then define our application deployments as "Plans". The "Plans" are defined by Kubenetes configuration yaml files. These config files can configure many different aspects like resource sizes, network settings, and health checks. These plans get communicated to our agent nodes for deployment on _all_ active machines. 
```

> We then define our application deployments as "Plans". The plans are defined by k8s configuration yaml files. These config files can configure many different aspects like resource sizes, network settings and health checks. These plans get communicated to our agent nodes for deployment on all active machines. 

Reviewer 2: ```suggestion
We then define our application deployments as "plans". The plans are defined by K8s configuration YAML files. These configuration files can configure many distinct aspects like resource sizes, network settings and health checks. These plans get communicated to our agent nodes for deployment on all active machines. 
```

![k8s-deployment](https://github.com/David-Elizondo/blog-usa/blob/master/images/2021/09/k8s3.png)

> Lastly, if we repeat our disaster scenario from before and one of our machines in our agent pool is taken offline. Our controllers now can assess the health of this application automatically and redeploy our application on the next available node. 

Reviewer: ```suggestion
Lastly, if we repeat our disaster scenario from before and one of our machines in our agent pool is taken offline, our controllers can now assess the health of this application automatically and redeploy our application onto the next available node.
```

> Lastly, if we repeat our disaster scenario from before and one of our machines in our agent pool is taken offline. Our controllers now can assess the health of this application automatically and redeploy our application on the next available node. 

Reviewer 2: ```suggestion
if we repeat our disaster scenario from before and one of our machines in our agent pool is taken offline, our controllers now can assess the health of this application automatically and redeploy our application on the next available node. 
```

![k8s-deployment](https://github.com/David-Elizondo/blog-usa/blob/master/images/2021/09/k8s4.png)

## Summary

> Using Kubernetes enables many new exciting features for Devops engineers. These functionalities are enabled by how we form our operational environment like I showed here. Using a cluster of machines we can achieve high availability. This change in deployment method allows our application deployments to withstand N machine failures where N is the size of your cluster.

Reviewer: ```suggestion
Using Kubernetes enables many new exciting features for us DevOps Engineers. These functionalities are enabled by how we form our operational environment like I demonstrated above. By using a cluster of machines, we can achieve high availability. This change in deployment method allows our application deployments to withstand x-number of machine failures where "y" is the size of your cluster.
```

> Using Kubernetes enables many new exciting features for Devops engineers. These functionalities are enabled by how we form our operational environment like I showed here. Using a cluster of machines we can achieve high availability. This change in deployment method allows our application deployments to withstand N machine failures where N is the size of your cluster.

Reviewer 2: ```suggestion
Using Kubernetes enables many exciting new features for DevOps engineers. These functionalities are enabled by how we form our operational environment like I showed here. Using a cluster of machines, we can achieve high availability. This change in deployment method allows our application deployments to withstand N machine failures where N is the size of your cluster.
```

> This high-level understanding of how clusters are formed and maintained should help your understanding of higher levels of complexities within larger and larger orchestration systems.

Reviewer: ```suggestion
This high level overview of how clusters are formed and maintained should contribute to your understanding of complexities within larger orchestration systems.
```

> This high-level understanding of how clusters are formed and maintained should help your understanding of higher levels of complexities within larger and larger orchestration systems.

Reviewer 2: Think about a call-to-action about how you or Ippon can help the reader with their services. Sell, sell, sell!

Author: Something like this? 

> Here at IPPON we have experience with designing resilient large scale application architectures. We do this by following the tech communities newest open source tools and best practices. If you are designing your next system or in charge of an Kubernetes cluster and in need of professional guidance send us over a message.

Reviewer 2: I just deleted a long wall of text. I am going to punt this to Youssra. She will be able to come up with something better than I can.

Author: Pushed!
